addons/controllers/cpi/cpiconfig_controller.go

```diff
@@ -0,0 +1,314 @@
+		return nil, err
+	}
+	d.VSphereCPI.Nsxt.Username, d.VSphereCPI.Nsxt.Password, err = getUsernameAndPasswordFromSecret(ctx, r.Client, nsxtSecret)
+	if err != nil {
+		return nil, err
+	}
+	d.VSphereCPI.Nsxt.Host = c.NSXT.Host
+	d.VSphereCPI.Nsxt.InsecureFlag = c.NSXT.InsecureFlag
+	d.VSphereCPI.Nsxt.RemoteAuth = c.NSXT.RemoteAuth
+	d.VSphereCPI.Nsxt.VmcAccessToken = c.NSXT.VMCAccessToken
+	d.VSphereCPI.Nsxt.VmcAuthHost = c.NSXT.VMCAuthHost
+	d.VSphereCPI.Nsxt.ClientCertKeyData = c.NSXT.ClientCertKeyData
+	d.VSphereCPI.Nsxt.ClientCertData = c.NSXT.ClientCertData
+	d.VSphereCPI.Nsxt.RootCAData = c.NSXT.RootCAData
+	d.VSphereCPI.Nsxt.SecretName = c.NSXT.SecretName
+	d.VSphereCPI.Nsxt.SecretNamespace = c.NSXT.SecretNamespace
+
+	d.VSphereCPI.HTTPProxy = c.HTTPProxy
+	d.VSphereCPI.HTTPSProxy = c.HTTPSProxy
+	d.VSphereCPI.NoProxy = c.NoProxy
+
+	return d, nil
+}
+
+func (r *CPIConfigReconciler) mapCPIConfigToDataValuesParavirtual(_ context.Context, cpiConfig *cpiv1alpha1.CPIConfig, _ *clusterapiv1beta1.Cluster) (*CPIDataValues, error) {
+	d := &CPIDataValues{}
+	c := cpiConfig.Spec.VSphereCPI
+
+	d.VSphereCPI.Mode = CPIParavirtualMode
+	d.VSphereCPI.ClusterAPIVersion = c.ClusterAPIVersion
+	d.VSphereCPI.ClusterKind = c.ClusterKind
+	d.VSphereCPI.ClusterName = c.ClusterName
+	d.VSphereCPI.ClusterUID = c.ClusterUID
+	d.VSphereCPI.SupervisorMasterEndpointIP = c.SupervisorMasterEndpointIP
+	d.VSphereCPI.SupervisorMasterPort = c.SupervisorMasterPort
+
+	return d, nil
+}
+
+// mapCPIConfigToDataValues maps CPIConfig CR to data values
+func (r *CPIConfigReconciler) mapCPIConfigToDataValues(ctx context.Context, cpiConfig *cpiv1alpha1.CPIConfig, cluster *clusterapiv1beta1.Cluster) (*CPIDataValues, error) {
+	switch cpiConfig.Spec.VSphereCPI.Mode {
+	case CPINonParavirtualMode:
+		return r.mapCPIConfigToDataValuesNonParavirtual(ctx, cpiConfig, cluster)
+	case CPIParavirtualMode:
+		return r.mapCPIConfigToDataValuesParavirtual(ctx, cpiConfig, cluster)
+	default:
+		break
+	}
+	return nil, errors.Errorf("Invalid CPI mode %s, must either be %s or %s", cpiConfig.Spec.VSphereCPI.Mode, CPIParavirtualMode, CPINonParavirtualMode)
+}
+
+// getOwnerCluster verifies that the CPIConfig has a cluster as its owner reference,
+// and returns the cluster
+func (r *CPIConfigReconciler) getOwnerCluster(ctx context.Context, cpiConfig *cpiv1alpha1.CPIConfig) (*clusterapiv1beta1.Cluster, error) {
+	cluster := &clusterapiv1beta1.Cluster{}
+	clusterName := cpiConfig.Name
+
+	for _, ownerRef := range cpiConfig.GetOwnerReferences() {
+		if ownerRef.Kind == "Cluster" {
+			clusterName = ownerRef.Name
+			break
+		}
+	}
+
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: cpiConfig.Namespace, Name: clusterName}, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info(fmt.Sprintf("Cluster resource '%s/%s' not found", cpiConfig.Namespace, clusterName))
+		}
+		r.Log.Error(err, fmt.Sprintf("Unable to fetch cluster '%s/%s'", cpiConfig.Namespace, clusterName))
```

Comment:
As you have made ```getOwnerCluster()``` a dedicated function and you are checking for nilness of the cluster here, you do not need to do the same check in the caller. You can return ```nil,nil``` for the case where the cluster is not found and simply check if the returned cluster object is nil in the caller. Otherwise you are duplicating this check in both places

Reply:
fixed